File: Assets/Dialogue/Scripts/Dialogue/CsharpOnly/DemoDialogueFunctions.cs
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

public class DemoDialogueFunctions : IDialogueFuncProvider
{
    [DialogueFunc("TestImmediateFunc")]
    public static void TestImmediateFunc(string param)
    {
        Debug.Log($"即时函数执行，参数: {param}");
    }
    
    [DialogueFunc("TestInteractiveFunc")]
    public static void TestInteractiveFunc(string param)
    {
        Debug.Log($"交互函数执行，参数: {param}");
    }
    
    [DialogueFunc("CheckCondition")]
    public static string CheckCondition(string branchA, string branchB)
    {
        // 简单条件判断，随机返回一个分支
        bool condition = Random.Range(0, 2) == 0;
        string result = condition ? branchA : branchB;
        Debug.Log($"条件判断，返回分支: {result}");
        return result;
    }
    
    [DialogueFunc("ShowSpecialEffect")]
    public static void ShowSpecialEffect(string effectName)
    {
        Debug.Log($"[Demo] 播放特殊特效：{effectName}");
    }

    [DialogueFunc("PlaySound")]
    public static void PlaySound(string soundName)
    {
        Debug.Log($"播放音效: {soundName}");
    }

    [DialogueFunc("StartDialogue")]
    public static void StartDialogue(string fileName)
    {
        Debug.Log($"启动新对话: {fileName}");
        DialogueController.Start(fileName);
    }

    [DialogueFunc("TestList")]
    public static void TestList(List<object> list)
    {
        Debug.Log($"[Demo] TestList: {StringUtil.FormatObject(list)}");
    }

    [DialogueFunc("TestDict")]
    public static void TestDict(Dictionary<object, object> dict)
    {
        Debug.Log($"[Demo] TestDict: {StringUtil.FormatObject(dict)}");
    }
}

File: Assets/Dialogue/LuaScripts/DialogueSystem/DemoDialogueFunctions.lua
--- 对话函数实现
--- 实现 IDialogueFuncProvider 接口以支持注册表扫描
local DemoDialogueFunctions = {}
local mt = {__index = require("DialogueFuncRegistry").IDialogueFuncProvider}
setmetatable(DemoDialogueFunctions, mt)

--- 规范定义：每个函数 = { name = "标识符", func = function(...) ... end }

--- 即时执行函数测试
DemoDialogueFunctions.TestImmediateFunc = {
    name = "TestImmediateFunc",
    func = function(param)
        CS.UnityEngine.Debug.Log("[Lua] 即时函数执行，参数: "..param)
    end
}

--- 交互执行函数测试
DemoDialogueFunctions.TestInteractiveFunc = {
    name = "TestInteractiveFunc",
    func = function(param)
        CS.UnityEngine.Debug.Log("[Lua] 交互函数执行，参数: "..param)
    end
} 

--- 条件判断函数测试
---@return string 选中的分支ID
DemoDialogueFunctions.CheckCondition = {
    name = "CheckCondition",
    paramTypes = {"string", "string"}, -- 指定参数类型为字符串
    func =  function(branchA, branchB)
        local condition = CS.UnityEngine.Random.Range(0, 2) == 0
        local result = condition and branchA or branchB
        result = branchB; -- 固定返回测试
        CS.UnityEngine.Debug.Log("[Lua] 条件判断，返回分支: "..result)
        return result
    end
}

--- 显示特效函数测试
DemoDialogueFunctions.ShowSpecialEffect = {
    name = "ShowSpecialEffect",
    func =  function(effectName)
        CS.UnityEngine.Debug.Log("[Lua] 显示特效: "..effectName)
    end
}

--- 播放音效函数测试
DemoDialogueFunctions.PlaySound = {
    name = "PlaySound",
    func =  function(soundName)
    CS.UnityEngine.Debug.Log("[Lua] 播放音效: "..soundName)
end
}

--- 启动新对话函数测试
DemoDialogueFunctions.StartDialogue = {
    name = "StartDialogue",
    func =  function(fileName)
        CS.UnityEngine.Debug.Log("[Lua] 启动新对话: "..fileName)
        local DialogueController = require("DialogueController")
        DialogueController.Start(fileName)
    end
}

---@function 辅助打印表
local stringUtil = require("StringUtil")

---@function 测试List参数
DemoDialogueFunctions.TestList = {
    name = "TestList",
    paramTypes = {"table"}, -- 指定参数类型为表（列表）
    func = function(list)
        CS.UnityEngine.Debug.Log("[Lua Demo] TestList: " .. stringUtil.Dump(list))
    end
}

---@function 测试Dict参数
DemoDialogueFunctions.TestDict = {
    name = "TestDict",
    func = function(dict)
        CS.UnityEngine.Debug.Log("[Lua Demo] TestDict: " .. stringUtil.Dump(dict))
    end
}

return DemoDialogueFunctions